File: demo_20_review/assignment_09.py
# -*- coding: utf-8 -*-
"""
##################################################
#
# QMB 6358: Software Tools for Business Analytics
#
# Sample Program for "Assignment 9"
#
# Lealand Morin, Ph.D.
# Assistant Professor
# Department of Economics
# College of Business Administration
# University of Central Florida
#
# December 1, 2021
#
##################################################
"""




##################################################
# Import Modules.
##################################################


import os # To set working directory
# import numpy as np # Not needed here but often useful
import pandas as pd # To read and inspect data
# from sklearn.linear_model import LinearRegression

import statsmodels.formula.api as sm # Another way to estimate linear regression



##################################################
# Set Working Directory.
##################################################



# Find out the current directory.
os.getcwd()
# Change to a new directory.
git_path = 'C:\\Users\\le279259\\OneDrive - University of Central Florida\\Documents\\Teaching\\QMB6358_Fall_2021\\GitRepo\\QMB6358F21\\'
os.chdir(git_path + 'demo_20_review')
# Check that the change was successful.
os.getcwd()



##################################################
# Load Data for Question 1.
##################################################


airplane_sales = pd.read_csv('airplane_sales.csv')


# Calculate summary statistics for your data.
airplane_sales.describe()



##################################################
# Linear Regression
##################################################



#--------------------------------------------------
# Fit the Regression Model (with statsmodels module instead).
#--------------------------------------------------

# Fit the regression model.
reg_model_full_sm = sm.ols(formula = "price ~ age", 
                           data = airplane_sales).fit()

# Display the parameters.
print(reg_model_full_sm.params)


# Display a summary table of regression results.
print(reg_model_full_sm.summary())



##################################################
# Load Data for Question 2.
##################################################


airplane_specs = pd.read_csv('airplane_specs.csv')


# Calculate summary statistics for your data.
airplane_specs.describe()


# Are the files in the same order?

airplane_sales.head()
airplane_specs.head()
airplane_sales.tail()
airplane_specs.tail()
# Not in same order.


# import sqlite3


# # Open the database from the previous session.
# con = sqlite3.connect('airplanes.db')
# # In this case, the database is already populated from the last demo.
# # Copy the file "population.db" to this folder to continue.

# # Create a cursor object from which to execute
# # SQL commands.
# cur = con.cursor()


# # Create another table to experiment with joins.
# cur.execute('''CREATE TABLE Sales(SaleID TEXT, Price REAL, Age REAL)''')
# # cur.execute('''DROP TABLE Sales''')


# airplane_sales_trim = pd.read_csv('airplane_sales_trim.csv')


# for row in airplane_sales_trim:
#  cur.execute('INSERT INTO Sales VALUES (?, ?, ?)', row)
# con.commit()



airplane_sales_sort = pd.read_csv('airplane_sales_sort.csv')
airplane_specs_sort = pd.read_csv('airplane_specs_sort.csv')




airplane_sales_specs = pd.concat([airplane_sales_sort, 
                                  airplane_specs_sort], 
                                 axis = 1)

airplane_sales_specs.describe()


airplane_sales_specs.rename(columns={'pass':'pass_2'}, 
                            inplace = True)


#--------------------------------------------------
# Fit the Regression Model (with statsmodels module instead).
#--------------------------------------------------

# Fit the regression model.
reg_model_full_sm = sm.ols(formula = "price ~ age + pass_2 + wtop + fixgear + tdrag", 
                           data = airplane_sales_specs).fit()

# Display the parameters.
print(reg_model_full_sm.params)


# Display a summary table of regression results.
print(reg_model_full_sm.summary())






##################################################
# Load Data for Question 3.
##################################################


airplane_perf_sort = pd.read_csv('airplane_perf_sort.csv')


# Calculate summary statistics for your data.
airplane_perf_sort.describe()




airplane_full = pd.concat([airplane_sales_specs, 
                           airplane_perf_sort], 
                          axis = 1)

airplane_full.describe()



#--------------------------------------------------
# Fit the Regression Model (with statsmodels module instead).
#--------------------------------------------------

# Fit the regression model.
reg_model_full_sm = sm.ols(formula = "price ~ age + pass_2 + wtop + fixgear + tdrag + horse + fuel + ceiling + cruise", 
                           data = airplane_full).fit()

# Display the parameters.
print(reg_model_full_sm.params)


# Display a summary table of regression results.
print(reg_model_full_sm.summary())




##################################################
# Optimize Function for Question 5.
##################################################


def ssr(beta, y, x1):
    return sum((y - beta[0] - beta[1]*x1)**2)


# Fit the regression model.
reg_model_full_sm = sm.ols(formula = "price ~ age", 
                           data = airplane_sales).fit()

# Display the parameters.
print(reg_model_full_sm.params)


# Display a summary table of regression results.
print(reg_model_full_sm.summary())




ssr(beta = reg_model_full_sm.params, 
    y = airplane_sales['price'], 
    x1 = airplane_sales['age'])

sum(reg_model_full_sm.resid**2)


import math
math.sqrt(ssr(beta = reg_model_full_sm.params, 
    y = airplane_sales['price'], 
    x1 = airplane_sales['age']))

y = airplane_sales['price']
x1 = airplane_sales['age']

import numpy as np
# Import scipy package for optimization
# from scipy import optimize
from scipy.optimize import minimize

beta_0 = np.zeros(len(reg_model_full_sm.params))
soln_bfgs = minimize(fun = ssr, x0 = beta_0, args = (y, x1), 
                     method = 'BFGS', 
                     options = {'xtol': 1e-8, 'maxiter': 1000, 'disp': True})


# The parameters:
print(soln_bfgs.x)
# Compare with the estimates from logit_model_fit_sm:
print(reg_model_full_sm.params)

File: demo_20_review/assignment_09.R

##################################################
#
# QMB 6358: Software Tools for Business Analytics
#
# Assignment 9
#
# Lealand Morin, Ph.D.
# Assistant Professor
# Department of Economics
# College of Business
# University of Central Florida
#
# December 1, 2021
#
##################################################



##################################################
# Preparing the Workspace
##################################################

# Clear workspace.
rm(list=ls(all=TRUE))

# Set working directory.
# wd_path <- '/path/to/your/folder'
wd_path <- 'C:/Users/le279259/OneDrive - University of Central Florida/Documents/Teaching/QMB6358_Fall_2021/GitRepo/QMB6358F21/demo_20_review'

setwd(wd_path)

# Check folder:
getwd()



##################################################
# Read the dataset and run regression for Question 1
##################################################

# Read the newly saved dataset.
airplane_sales <- read.csv(file = 'airplane_sales.csv')

# Inspect the data.
summary(airplane_sales)


##################################################
# Estimating the Regression Model
# Model 1: All Variables Included
##################################################


# Estimate a regression model.
lm_model_1 <- lm(data = airplane_sales,
                    formula = price ~ age)

# Output the results to screen.
summary(lm_model_1)


##################################################
# Read the dataset and run regression for Question 2
##################################################

# Read the newly saved dataset.
airplane_specs <- read.csv(file = 'airplane_specs.csv')

# Inspect the data.
summary(airplane_specs)


airplane_sales_specs <- merge(x = airplane_sales,
                              y = airplane_specs,
                              by = 'X0Sale_ID')


summary(airplane_sales_specs)
colnames(airplane_sales_specs)

# Estimate a regression model.
lm_model_2 <- lm(data = airplane_sales_specs,
                 formula = price ~ age +
                 pass + wtop + fixgear + tdrag)

# Output the results to screen.
summary(lm_model_2)




##################################################
# Read the dataset and run regression for Question 3
##################################################

# Read the newly saved dataset.
airplane_perf <- read.csv(file = 'airplane_perf.csv')

# Inspect the data.
summary(airplane_perf)


airplane_full <- merge(x = airplane_sales_specs,
                       y = airplane_perf,
                       by = 'X0Sale_ID')


summary(airplane_full)

# Estimate a regression model.
lm_model_3 <- lm(data = airplane_full,
                 formula = price ~ age +
                   pass + wtop + fixgear + tdrag +
                   horse + fuel + ceiling + cruise )

# Output the results to screen.
summary(lm_model_3)



colnames(airplane_full)


log_list <- colnames(airplane_full)[c(3,2,8,9,10,11)]
for (var_name in log_list) {
  log_var_name <- sprintf('log_%s', var_name)
  airplane_full[, log_var_name] <- log(airplane_full[, var_name])
}

summary(airplane_full)



# Estimate a regression model.
lm_model_4 <- lm(data = airplane_full,
                 formula = log_price ~ log_age +
                   pass + wtop + fixgear + tdrag +
                   log_horse + log_fuel + log_ceiling + log_cruise)

# Output the results to screen.
summary(lm_model_4)



##################################################
# Question 5
##################################################

ssr <- function(beta, y, x) {
  return(sum((y - beta[1] - beta[2]*x)^2))
}



# Output the results to screen.
summary(lm_model_1)

sum(lm_model_1$residuals^2)

coef(lm_model_1)[1]
coef(lm_model_1)[2]

# Test the function to math the regression summary.
# beta_test_ssr <- c(coef(lm_model_1)['(Intercept)'],
#                    coef(lm_model_1)['age'])


sum(lm_model_1$residuals^2)

ssr(beta = coef(lm_model_1),
    y = airplane_full[, 'price'],
    x = airplane_full[, 'age'])

# Optimize the ssr function to obtain the estimates.
soln_ssr <- optim(c(0, 0), ssr,
                  y = airplane_full[, 'price'],
                  x = airplane_full[, 'age'])


print(soln_ssr)


# This is cheating:

# Optimize the ssr function to obtain the estimates.
soln_ssr <- optim(coef(lm_model_1), ssr,
                  y = airplane_full[, 'price'],
                  x = airplane_full[, 'age'])


print(soln_ssr)
# Closer but this is cheating by using the answer: coef(lm_model_1)

